---
title: '*dms-viz*: Structure-informed visualizations for deep mutational scanning and other mutation-based datasets'
papersize: a4
tags:
  - javascript
  - data visualization
  - molecular biology
  - protein biology
  - deep mutational scanning
authors:
  - name: William W. Hannon
    orcid: 0000-0002-6014-4749
    affiliation: "1, 2" 
  - name: Jesse D. Bloom
    orcid: 0000-0003-1267-3408
    affiliation: "2, 3, 4" 
affiliations:
 - name: Molecular and Cellular Biology Graduate Program, University of Washington, Seattle, Washington 98109, United States of America
   index: 1
 - name: Basic Sciences and Computational Biology, Fred Hutchinson Cancer Research Center, Seattle, Washington 98109, United States of America
   index: 2
 - name: Department of Genome Sciences, University of Washington, Seattle, Washington 98109, United States of America
   index: 3
 - name: Howard Hughes Medical Institute, Seattle, Washington 98109, United States of America
   index: 4
date: 26 October 2023
bibliography: paper.bib
---

# Summary and Purpose

Many biological questions require an understanding of how mutations impact a protein's functions. Deep mutational scanning (DMS) offers an approach to characterize the impact of a huge number of mutations in parallel [@fowlerDeepMutationalScanning2014]. The wide application of DMS has greatly increased the number of mutation-function datasets [@fowlerAtlasVariantEffects2023]. For instance, DMS has been used to determine how mutations to viral proteins affect antibody escape [@dadonaitePseudovirusSystemEnables2023], receptor affinity [@starrDeepMutationalScanning2020], and essential functions such as viral genome transcription and replication [@liDeepMutationalScanning2023]. In some cases, the effects of mutations can also be inferred from phylogenies of natural sequences [@bloomFitnessEffectsMutations2023] (\autoref{fig:figure1}).

![Large mutation-associated datasets can be used to map antibody footprints, assess the impact of mutations on protein function, and identify patterns of selection from natural mutation frequencies.\label{fig:figure1}](figures/figure-1.png)

The mutation-based data generated by these approaches is better understood in the context of a protein’s 3D structure. However, current approaches for visualizing mutation data in the context of a protein’s structure are often cumbersome and require multiple steps and software. To streamline the visualization of mutation-associated data in the context of a protein structure, we developed a web-based tool, `dms-viz`. With `dms-viz`, users can straightforwardly visualize mutation-based data such as those from DMS experiments in the context of a 3D protein model in an interactive format. Visit <https://dms-viz.github.io/> to use `dms-viz`.

# Statement of Need

We wanted `dms-viz` to provide the following functionalities:

1. **Provide structural context**: `dms-viz` simplifies the process of visualizing mutation data with structural context by superimposing mutation measurements on a 3D protein structure. Additionally, it provides extensive control over the visual representation of the 3D structure.

2. **Accommodate diverse data types**: Although analyzing DMS data is a key goal of `dms-viz`, there are many types of mutation data. The tool can handle diverse data types via a command line interface that converts data into a common format.

3. **Display multiple conditions**: With `dms-viz`, multiple experimental conditions can be visualized concurrently; for instance, researchers can easily visualize multiple antibody binding footprints from polyclonal sera [@yuBiophysicalModelViral2022].

4. **Maximize customization**: Every dataset has specific needs for visual representation. Recognizing this, `dms-viz` offers customization with filters (which are important for navigating large and possibly noisy datasets), and tooltips, ensuring that nuances are communicated.

5. **Create compact interactive visualizations**: `dms-viz` creates compact interactive views that can be incorporated into HTML presentation slides (e.g., <https://slides.com/>).

6. **Share findings with ease**: Users of `dms-viz` can generate shareable URL links to their visualizations. They can also save and share their `JSON` specification files, ensuring that data can be accessed by others.

7. **Preserve data privacy**: `dms-viz` allows users to visualize proprietary structures and analyze sensitive data in their browser without uploading their datasets to a remote server or storing them in a public repository.

Our group previously created a tool called `dms-view` [@hiltonDmsviewInteractiveVisualization2020] that has some of the functionalities listed above. However, we designed `dms-viz` to be more customizable and comprehensive to handle a wider diversity of experimental designs and questions.

# Design and Usage

Using `dms-viz` involves three components. First, using the command line tool `configure-dms-viz`, available as a Python package on PyPI (<https://pypi.org/project/configure-dms-viz/>), the user formats their data into a `JSON` specification file (see the [documentation](https://dms-viz.github.io/dms-viz-docs/preparing-data/command-line-api/#json-schema) for details on the `JSON` schema). Then, the user uploads this specification file to `dms-viz.github.io`, a web-based interface written in Javascript, `D3.js`, and `NGL.js` [@roseNGLViewerWebbased2018]. Finally, the specification file can either be shared directly or hosted remotely to generate a shareable URL link (\autoref{fig:figure2}).

![(1) The user formats their data using the command line tool `configure-dms-viz`. (2) The user takes the resulting `JSON` specification file and uploads it to `dms-viz.github.io`. (3) The user shares their results with a `JSON` file, a URL link, or a static image.\label{fig:figure2}](figures/figure-2.png)

Upon uploading the specification file to `dms-viz`, users will see a visualization composed of four components, as illustrated in \autoref{fig:figure3}.

1. ***Context* plot**: Located at the top of the visualization, it allows users to zoom into specific sites on the *Focus* plot while maintaining an overview of the entire dataset.

2. ***Focus* plot**: This plot shows a summarized view of the user's data. Every measured protein site is represented as a point providing a summary statistic of the effects of mutations at that site, and adjacent sites are connected with lines.

3. ***Detail* heatmap**: If the user is interested in the measurements for every mutation at a site, they can click on that site in the *Focus* plot. This will populate a heatmap with each mutation measurement at that site.

4. ***Interactive* structure**: When the user wants structural context for a given set of sites, they can drag a brush over the corresponding points in the *Focus* plot. This action will highlight those sites on an interactive 3D protein model.

To ensure the visualization remains compact, all configuration options are tucked away in a collapsible sidebar. See the documentation at <https://dms-viz.github.io/dms-viz-docs/> for more information about how to use `dms-viz` along with detailed tutorials and examples.

![`dms-viz` provides a compact interface for exploring mutation-associated data, in this case, mutation-escape from the constituents of a therapeutic antibody cocktail measured by DMS of the SARS-CoV-2 receptor binding domain (RBD) [@starrProspectiveMappingViral2021]. In this example, the structure shown is the SARS-CoV-2 RBD bound to both antibodies in the therapeutic cocktail (PDB: 6XDG).\label{fig:figure3}](figures/figure-3.png)

# Examples

For additional scientific background and a walkthrough of the code that generates these visualizations, visit the [documentation](https://dms-viz.github.io/dms-viz-docs/visualizing-data/vignettes/).

## 1. Mapping the neutralization profile of antibodies and sera against HIV envelope

@radfordMappingNeutralizingSpecificity2023 mapped mutations to HIV envelope (Env) that affect neutralization by polyclonal human serum using a pseudotyping-based deep mutational scanning platform [@radfordMappingNeutralizingSpecificity2023]. See how `dms-viz` can be used to interactively visualize datasets with multiple antibody footprints on a single summary plot [here](https://dms-viz.github.io/v0/?data=https%3A%2F%2Fraw.githubusercontent.com%2Fdms-viz%2Fconfigure_dms_viz%2Fmain%2Ftests%2FHIV-Envelope-BF520-DMS%2Foutput%2FHIV-Envelope-BF520-DMS.json&e=IDC508&ce=%255B%25221%2522%252C%25222%2522%255D).

## 2. Using mutation-fitness data to augment structure-guided drug design

Bloom and Neher estimated the fitness effects of mutations to all SARS-CoV-2 proteins by analyzing millions of human SARS-CoV-2 sequences  [@bloomFitnessEffectsMutations2023]. See how `dms-viz` can be used to enhance structure-guided drug design by merging this data with structural views of a viral target like the SARS-CoV-2 main protease (Mpro) in complex with a bound ligand such as MAT-POS-e194df51-1 from the COVID Moonshot project [@bobyOpenScienceDiscovery2023] [here](https://dms-viz.github.io/v0/?data=https%3A%2F%2Fgist.githubusercontent.com%2FWillHannon-MCB%2Faa8599278999c841550c54f12b9964ee%2Fraw%2Fea7540c9b8ce0795dbbc9bf60cae85ba37cc7635%2Fmpro_with_inhibitor.json&markdown=https%3A%2F%2Fgist.githubusercontent.com%2FWillHannon-MCB%2Faa8599278999c841550c54f12b9964ee%2Fraw%2F2984be0aa286cf6a3447f65230b2ae9b7d128671%2FREADME.md&g=true&le=true&fi=%257B%2522expected_count%2522%253A10.15084%257D&sa=true&so=0.58&lr=ball%2Bstick&pr=ball%2Bstick&sr=surface).

## 3. Visualizing the pathogenicity of genetic variants of an important tumor suppressor

@matreyekMultiplexAssessmentProtein2018 used variant abundance by massively parallel sequencing (VAMP-seq) to characterize the effect of thousands of mutations on the intracellular abundance of PTEN, a tumor suppressor that is inactivated in many cancers [@matreyekMultiplexAssessmentProtein2018]. See how `dms-viz` can be used to identify clinically relevant mutations in human proteins [here](https://dms-viz.github.io/v0/?data=https%3A%2F%2Fgist.githubusercontent.com%2FWillHannon-MCB%2F102f0e3dff8f67452ffe3f9f290c41cf%2Fraw%2F7df95064f74c0f9681cd5fec883306aa5a021232%2Fpten_example.json).

# Conclusion

`dms-viz` is a valuable addition to the suite of computational tools available for analyzing, sharing, and visualizing mutation-based datasets, which includes [MaveDB](https://www.mavedb.org/), [ProtVar](https://www.ebi.ac.uk/ProtVar/about), and many others [@espositoMaveDBOpensourcePlatform2019]. We designed `dms-viz`  as a practical and user-friendly approach to visualizing mutation-associated data in the context of protein structures.  Because `dms-viz` is capable of handling various data types and has options for both sharing and privacy, it should apply to the visualization of a wide range of datasets.

# Code Availability

- The visualization is available at <https://dms-viz.github.io/>
- The documentation is available at <https://dms-viz.github.io/dms-viz-docs/>
- The source code for `dms-viz.github.io` is available at <https://github.com/dms-viz/dms-viz.github.io>
- The source code for `configure-dms-viz` is available at <https://github.com/dms-viz/configure_dms_viz>

# Acknowledgements

This project was envisioned as the successor to `dms-view`. Thank you to Dr. Sarah Hilton and Dr. John Huddleston for laying this groundwork and for providing helpful input. Thank you to the Bloom lab for providing data and guidance. The research reported here was supported in part by NIAID of the National Institutes of Health under award number U19AI171399. The content is solely the responsibility of the authors and does not necessarily represent the official views of the National Institutes of Health. The work of JDB was supported in part by the NIH/NIAID under grants R01AI141707 and contract 75N93021C00015. JDB is an Investigator of the Howard Hughes Medical Institute.

# Disclosures

JDB is on the scientific advisory boards of Apriori Bio, Aerium Therapeutics, Invivyd, and the Vaccine Company. JDB receives royalty payments as an inventor on Fred Hutch licensed patents related to deep mutational scanning of viral proteins.

# References
